### ISSUE TITLE
no-invalid-this: false positive with `this` in class field

### ISSUE THREAD

**Issue by regseb** [eslint/eslint#14221]
**Tell us about your environment**

* **ESLint Version:** 7.22.0
* **Node Version:** v14.16.0
* **npm Version:** 7.6.1
* **Operating System:** Ubuntu 20.04.2 LTS

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?** `@babel/eslint-parser`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
    parser: "@babel/eslint-parser",
    parserOptions: {
        requireConfigFile: false,
        babelOptions: {
            plugins: ["@babel/plugin-proposal-class-properties"],
        },
    },

    rules: {
        "no-invalid-this": 2,
    },
};
```

</details>

<details>
<summary>package.json</summary>

```json
{
    "name": "testcase",
    "version": "1.0.0",
    "dependencies": {
        "@babel/core": "^7.13.10",
        "@babel/eslint-parser": "^7.13.10",
        "@babel/plugin-proposal-class-properties": "^7.13.0",
        "eslint": "^7.22.0"
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
class Foo {
    bar = this;
}
```

```bash
npx eslint index.js
```

**What did you expect to happen?** No error reported.

**What actually happened? Please copy-paste the actual, raw output from ESLint.**

> /home/regseb/testcase/index.js
>   2:11  error  Unexpected 'this'  no-invalid-this
>
> ✖ 1 problem (1 error, 0 warnings)

**Steps to reproduce this issue:**

1. npm i
1. npx eslint index.js

**Are you willing to submit a pull request to fix this bug?** No.

**Comment by ljharb** [comment]
Class fields isn't stage 4, so despite it shipping in most engines, eslint doesn't yet support it. Since the rule doesn't crash, I'm not sure they'll fix it until the proposal advances.

If not, then you'd want to use `@babel/eslint-plugin` and the `@babel/no-invalid-this` rule.

**Comment by mdjermanovic** [comment]
Hi @regseb, thanks for the issue!

As @ljharb explained, this rule doesn't support stage 3 class syntax. Per our [policy around experimental features](https://github.com/eslint/eslint#what-about-experimental-features), we'll update ESLint's default parser and all core rules if and when this syntax reaches stage 4.

In the meantime, you can use [@babel/eslint-plugin](https://www.npmjs.com/package/@babel/eslint-plugin), which provides a replacement rule `@babel/no-invalid-this`:

> `@babel/no-invalid-this`: handles class fields and private class methods (`class A { a = this.b; }`)

I'm closing this as it isn't actionable at the moment, and a replacement rule already exists in a plugin.
